### PR TITLE
feature: list of licenses

### LINKED ISSUE #5575: Multiple licenses
## Desired Behavior
Allow `license` field to take multiple arguments.

## Example
`(license MIT LGPL-2.1-only)`

### PR DESCRIPTION
**Pull request by liyishuai**
Resolve #5575

### CODE REVIEW

**Comment by rgrinberg** [comment]
Thanks for your contribution. When introducing features in dune, we try to preserve the old behavior unless the user explicitly upgrades the version in the dune-project file. Therefore, we need to update the PR to:

1. Reject multiple licenses unless language version of dune is at least 3.2
2. Generate singular license fields pre 3.2

Also, for 3.2 and above, I would still prefer if the license field was singular if there was only one license specified when generating the opam file.

If you have any questions on how to implement this, please ask away.

File: src/dune_rules/opam_create.ml [hunk 1]
```diff
@@ -227,6 +230,10 @@ let opam_fields project (package : Package.t) =
   let list_fields =
     [ ("maintainer", Package.Info.maintainers info)
     ; ("authors", Package.Info.authors info)
+    ; ( "license"
+      , match Package.Info.license info with
+        | None | Some [ _ ] -> None
+        | Some l -> Some l )
```
Comment (liyishuai) [inline src/dune_rules/opam_create.ml:236]: This includes `[]` and `_::_::_`. Should we reject empty license list as well?

Comment (rgrinberg) [reply src/dune_rules/opam_create.ml:236]: I'd reject the empty list since it wasn't allowed before.

File: .gitignore [hunk 2]
```diff
@@ -17,6 +17,6 @@ _perf
 /boot/*.cm*
 dune.exe
 Makefile.dev
-src/dune/setup.ml
+src/dune_rules/setup.ml
```
Comment (rgrinberg) [inline .gitignore:20]: Could you move this into a separate commit? If not, I could do so myself as well.

File: src/dune_rules/opam_create.ml (hunk 1 above)
Comment (liyishuai) [reply src/dune_rules/opam_create.ml:236]: Now rejected in parsing phase.

**Comment by rgrinberg** [comment]
Thanks. Could you add some tests cases? See dune-project-meta/main.t for example test cases.

**Comment by kit-ty-kate** [comment]
Note that, while correct, opam 2.2.0 and above will come with a proper SPDX parser (https://github.com/ocaml/opam/pull/4768), which allows the `<license> AND <license>` and `<license> OR <license>` syntaxes.

File: src/dune_engine/package.ml [hunk 3]
```diff
@@ -480,7 +480,7 @@ module Info = struct
       [ field_o "source" Source_kind.encode source
       ; field_l "authors" string (Option.value ~default:[] authors)
       ; field_l "maintainers" string (Option.value ~default:[] maintainers)
-      ; field_o "license" string license
+      ; field_l "license" string (Option.value ~default:[] license)
```
Comment (kit-ty-kate) [inline src/dune_engine/package.ml:483]: What does this do in the presence of say `(license LGPL-2.1-or-later WITH OCaml-LGPL-linking-exeption)` ?

Comment (liyishuai) [reply src/dune_engine/package.ml:483]: It'll yield `license: ["LGPL-2.1-or-later" "WITH" "OCaml-LGPL-linking-exeption"]`.

Is `WITH` currently handled somewhere?

Comment (kit-ty-kate) [reply src/dune_engine/package.ml:483]: in opam yes, this is very common. It should be outputed as one string as before

Comment (liyishuai) [reply src/dune_engine/package.ml:483]: Did you mean `(license "LGPL-2.1-or-later WITH OCaml-LGPL-linking-exeption")`?
Space-separated list was rejected before 3.2.

Comment (kit-ty-kate) [reply src/dune_engine/package.ml:483]: ah i see, yes. Then it should be fine.

Comment (rgrinberg) [reply src/dune_engine/package.ml:483]: An alternative to this would  be to introduce a `licenses` stanza instead. This similar as to how we have `executable` and `executables` stanzas for example, and it would avoid this gotcha.

@kit-ty-kate @liyishuai what do you think?

Comment (kit-ty-kate) [reply src/dune_engine/package.ml:483]: i think it’s fine with just one as it reflects the opam field of the same name

Comment (rgrinberg) [reply src/dune_engine/package.ml:483]: Okay, but isn't this going to break compat for people who have:

```
(lang dune 3.1)
(license LGPL-2.1-or-later WITH OCaml-LGPL-linking-exeption)
```

And they will upgrade to `3.2`?

Comment (kit-ty-kate) [reply src/dune_engine/package.ml:483]: no, as @liyishuai said this dune-project file would result in an error:
```
$ cat dune-project
(lang dune 3.1)
(license LGPL-2.1-or-later WITH OCaml-LGPL-linking-exeption)
$ dune build
File "dune-project", line 2, characters 27-31:
2 | (license LGPL-2.1-or-later WITH OCaml-LGPL-linking-exeption)
                               ^^^^
Error: Too many argument for license
```

**Comment by rgrinberg** [comment]
Merged. Thanks.
